### PR TITLE
Don't attempt to default-construct a reference.

### PR DESCRIPTION
**Pull request by BillyONeal**
Currently, Expected<T> has a bug for references, in that it is default constructible, but references shouldn't be default constructible. (This is hidden because we store a pointer internally).

In VersionedPackageGraph::require_port_version, an ExpectedS<T&> is default constructed, which is in the "impossible" state of neither having a stored T (since references can't be default constructed), but also not being assigned an error state.

This change hoists the only place where the ExpectedS could contain an error into an early-return, allowing a plain pointer to be used, and avoiding constructing the impossible thing.

(The bug in Expected that allowed this to happen is fixed in https://github.com/microsoft/vcpkg-tool/pull/511 , but this change was tricky enough I thought extracting it for review alone made sense)

Reasoning, using "after" line numbers:

1557-1567: This if block could never have stored an error state in the expected, since it is only entered if there is an overlay control file that will be the result. Therefore, we will never add an error to m_errors from here.
1578: get_control_file can return an expected in the error state, so we have to check if it's an error and, if so, add it to m_errors.
1587: At this point, we know p_scfl is not nullptr since both if branches set it on 1556 or 1578.

### CODE REVIEW

File: src/vcpkg/dependencies.cpp [hunk 1]
```diff
@@ -1551,79 +1551,79 @@ namespace vcpkg::Dependencies
...
-                    {
-                        for (auto&& f : p_scfl->source_control_file->core_paragraph->default_features)
-                        {
-                            add_feature_to(graph_entry, versioned_graph_entry, f);
-                        }
-                    }
                 }
             }
-            else
-            {
```
Comment (BillyONeal) [inline src/vcpkg/dependencies.cpp:1624]: Note that this block got (effectively) moved to 1580

File: src/vcpkg/dependencies.cpp [hunk 2]
```diff
@@ -1551,79 +1551,79 @@ namespace vcpkg::Dependencies
...
-                {
-                    replace = true;
-                }
-                else if (versioned_graph_entry.scfl == p_scfl)
+                const auto maybe_scfl = m_ver_provider.get_control_file({graph_entry.first.name(), version});
+                p_scfl = maybe_scfl.get();
+                if (!p_scfl)
                 {
-                    replace = false;
+                    m_errors.push_back(maybe_scfl.error());
```
Comment (ras0219-msft) [inline src/vcpkg/dependencies.cpp:1582]: This could be a move if `maybe_scfl` wasn't const.

**Review by ras0219-msft — APPROVED** [review]

File: src/vcpkg/dependencies.cpp [hunk 3]
```diff
@@ -1551,79 +1551,79 @@ namespace vcpkg::Dependencies
...
+                    return;
                 }
-                else
+            }
+
+            auto& versioned_graph_entry = graph_entry.second.emplace_node(
```
[the remark this answers is not in the document]
Comment (ras0219-msft) [reply src/vcpkg/dependencies.cpp]: If you toggle "ignore whitespace", it recognizes that nothing changed.

[the remark this answers is not in the document]
Comment (BillyONeal) [reply src/vcpkg/dependencies.cpp]: :sigh: The alignment is still bad

File: src/vcpkg/dependencies.cpp (hunk 2 above)
Comment (BillyONeal) [reply src/vcpkg/dependencies.cpp]: Fixed!

Comment (dg0yt) [reply src/vcpkg/dependencies.cpp]: But (now) it still pushes the error, not the "moved" object?

Comment (BillyONeal) [reply src/vcpkg/dependencies.cpp]: @dg0yt: It's still pushing a moved object; that calls


https://github.com/microsoft/vcpkg-tool/blob/01294638b93e0d472007491d449db1ab7cdf3aa4/include/vcpkg/base/expected.h#L182

**Pull request merged**